virt-launcher, migration: Protect the setup using a lock, fix some missing operations and refactor the code

EdDev opened the pull request:
**What this PR does / why we need it**:

Operations on the domain which involve its state change are protected
from concurrent access using a domain-modify-lock (mutex).

In the migration setup stage, currently two operations are affecting the
domain state:
- Detaching of SR-IOV devices.
- Preparing migration parameters (which includes the domxml)

These operation need to be protected from change of the domain and
therefore the domain-modify-lock is added.

As part of this journey, a refactoring was needed to avoid double-locking.
In addition, some small fixes have been added:
- Report migration status on all failures.
- Abort the detach operation in case of failure.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
In order to follow the changes, it is recommended to review this change one commit at a time.

**Release note**:
```release-note
NONE
```

vladikr commented:
This change makes sense to me.
Let's see that all tests pass.

EdDev commented:
/cc @AlonaKaplan

vladikr commented:
/approve

davidvossel commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -580,6 +538,90 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
+	for _, port := range migrationPortsRange {
+		key := migrationproxy.ConstructProxyKey(string(vmi.UID), port)
+		migrationProxy := migrationproxy.NewTargetProxy(loopbackAddress, port, nil, nil, migrationproxy.SourceUnixFile(virtShareDir, key))
+		err := migrationProxy.StartListening()
+		if err != nil {
+			return stopServer, err
+		}
```
`err` is defined as a return argument in the function. Here you're defining `err` again within the scope of this for loop. In the defer function you're checking 

```
		if err != nil {
			stopServer()
			stopServer = nil
		}
```

due to the redefining of `err` at different scopes, it's unclear to me if the logic in that defer loop will do what you'd expect it to do when an error is encountered.

Given the complexity here, maybe this particular block shouldn't be refactored, and was easier to maintain in the top level function. Is that possible?

vladikr replied:
I thought that this doesn't matter because we explicitly return the err 
But I understand that this may be confusing.

EdDev replied:
> due to the redefining of `err` at different scopes, it's unclear to me if the logic in that defer loop will do what you'd expect it to do when an error is encountered.
> 
> Given the complexity here, maybe this particular block shouldn't be refactored, and was easier to maintain in the top level function. Is that possible?

I recognize that reusing the same name may be confusing, I can use a simple `e` in this case if it helps.
Regarding the complexity, I think this is a common pattern when using defers and it is used in the codebase especially when the deferred operation returns an error (like the famous file close).

The change itself started when I still did not know how the end result will look like, eventually this portion was not directly needed for the locking of the domain change, so we could remove it completely from this PR.

I can argue the following in its favor:
- The abstraction this change introduces is simplifying the caller significantly, but indeed adds some complexity in the helper.
- The original usage of the `defer` inside the for-loop is not less complex, it adds a "order" question mark (my IDE warns about using defers in loops).
- The usage of a named returned value is part of golang and is used on the codebase already.

Summarizing, if changing the name of the error variable will help, I would gladly favor that over dropping this change.
However, if you feel it is too much, I will drop the change.
Please let me know your preference on this.

davidvossel replied:
personally, the indirection in this helper function takes longer for me to grok than seeing everything laid out sequentially in the caller function. not overloading the err variable would make it more obvious. returning the logic to the caller would make it more obvious too. whatever you want to do.

EdDev replied:
Thank you for freedom to choose one way or the other.
I prepared both options and went back and forth between them. I feel more comfortable with abstracting the layer of the proxy preparation, allowing the reader to choose if to delve into the details (about the how) or not.

I changed the error variable name and scoped it in the `if` statement.

davidvossel reviewed (commented):
looks good. i have some questions about one part

EdDev commented:
change: rebase

EdDev commented:
[change](https://github.com/kubevirt/kubevirt/compare/31bb04dc273f9b204a4d68675d985cf3efb32547..79a021d441329012a042e887b77411459f8d6745):
- Renamed the error variable and scoped it in the `if` statement to make it less confusing (with the returned named one).
- Moved the comment about the proxy to the function.

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -580,6 +558,50 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
+	if e := libvirtConnectionProxy.StartListening(); e != nil {
+		return stopServer, e
+	}
+	proxyStopListenPool = append(proxyStopListenPool, libvirtConnectionProxy.StopListening)
```
I have to admit for me it was very hard to understand the logic of  `proxyStopListenPool` comparing to the previous defer logic.
A method for stopping the server that invokes a pool of stopping methods which is invoked by one of the two defers, depends whether there was an error or not.
I felt like I'm solving an interview question:)
Can you please explain the added value of this. Readability is for sure not one of the added values.

AlonaKaplan replied:
> I found no other way to implement the needed. If you have something else in mind, please let me know.
Let the caller do the clean in case of an error.
Anyway we depend on the caller doing the cleanup if there is no error.

You can leave it as is, I just find it very confusing for the "future developer".

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -580,6 +558,50 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
+		key := migrationproxy.ConstructProxyKey(string(vmi.UID), port)
+		migrationProxy := migrationproxy.NewTargetProxy(loopbackAddress, port, nil, nil, migrationproxy.SourceUnixFile(virtShareDir, key))
+		if e := migrationProxy.StartListening(); e != nil {
+			return stopServer, e
```
Previously the `StopListening` was invoked even in case of an error. Are you sure the is nothing to clean/stop in case of an error?
If so, please updated the commit message that the commit is not just and extract of a method.

AlonaKaplan replied:
I mean that in case of a error, you return before appending the `StopListening` method to the `proxyStopListenPool`.
It means `StopListening` won't be called on a error migration proxy.
Before this commit, it was invoked also in case of an error.

EdDev replied:
Yes, if an error occurs, the expectation is that there is nothing to cleanup.
Looking into the existing function, the error may occur only to create the socket, so I could not see any issue here.

I will update the commit message.

EdDev replied:
Done

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -580,6 +558,50 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
+	//  proxy incoming migration requests on port 22222 to the vmi's existing libvirt connection
+	libvirtConnectionProxy := migrationproxy.NewTargetProxy(loopbackAddress, LibvirtLocalConnectionPort, nil, nil, migrationproxy.SourceUnixFile(virtShareDir, string(vmi.UID)))
+	if e := libvirtConnectionProxy.StartListening(); e != nil {
+		return stopServer, e
```
same

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -531,6 +531,9 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
 }
 
 func setupMigration(l *LibvirtDomainManager, vmi *v1.VirtualMachineInstance, dom cli.VirDomain, options *cmdclient.MigrationOptions, loopbackAddress string) (*libvirt.DomainMigrateParameters, error) {
+	l.domainModifyLock.Lock()
```
Can you please explain in the commit message why `l.virConn.LookupDomainByName` is not part of the lock (as in other places)?

EdDev replied:
I tried to avoid it in order not to admit we are not solving everything with this.
I am adding the explanation.

AlonaKaplan commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -531,6 +531,9 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
 
 func setupMigration(l *LibvirtDomainManager, vmi *v1.VirtualMachineInstance, dom cli.VirDomain, options *cmdclient.MigrationOptions, loopbackAddress string) (*libvirt.DomainMigrateParameters, error) {
+	l.domainModifyLock.Lock()
+	defer l.domainModifyLock.Unlock()
```
Why `getDiskTargetsForMigration` is not part of the lock? Seems it reads the domain as well.
Same for `MigrateToURI3`.

EdDev replied:
> Why `getDiskTargetsForMigration` is not part of the lock? Seems it reads the domain as well.

It is included in the lock.

> Same for `MigrateToURI3`.

I will explain in the commit message.
We cannot do it because of the go-routine that monitors the migration progress, it uses the same lock.

> Why not looking before `LookupDomainByName` and releasing the lock on the `defer` of the go routine?

It is unsafe to lock in one place and release in another (a different abstraction layer, different go-routine) and I think it would not have worked in this specific case (see the answer before about the monitor go-routine).

EdDev commented:
change: Rebase

EdDev commented:
> change: Rebase

Ahh, will need this again as the PR which required this will get reverted soon.

AlonaKaplan commented:
/lgtm

EdDev commented:
change: Rebase

enp0s3 reviewed (approved):
/lgtm

AlonaKaplan commented:
/lgtm

EdDev commented:
/retest

EdDev commented:
/retest

EdDev commented:
/retest

EdDev commented:
/retest
